server/src/main/java/org/opensearch/action/admin/cluster/remotestore/repository/RemoteStoreRepositoryRegistrationHelper.java

```diff
@@ -0,0 +1,287 @@
+            throw new IllegalStateException(
+                "joining node ["
+                    + joiningNode
+                    + "] has node attribute ["
+                    + REMOTE_STORE_TRANSLOG_REPOSITORY_SETTINGS_ATTRIBUTE_KEY
+                    + "] different than existing node ["
+                    + existingNode
+                    + "]"
+            );
+        }
+    }
+
+    private static Settings buildSettings(String stringSettings) {
+        Settings.Builder settings = Settings.builder();
+
+        String[] stringKeyValue = stringSettings.split(",");
+        for (int i = 0; i < stringKeyValue.length; i++) {
+            String[] keyValue = stringKeyValue[i].split(":");
+            settings.put(keyValue[0].trim(), keyValue[1].trim());
+        }
+
+        return settings.build();
+    }
+
+    private static RepositoryMetadata buildSegmentRepositoryMetadata(DiscoveryNode node) {
+        String name = node.getAttributes().get(REMOTE_STORE_SEGMENT_REPOSITORY_NAME_ATTRIBUTE_KEY);
+        String type = node.getAttributes().get(REMOTE_STORE_SEGMENT_REPOSITORY_TYPE_ATTRIBUTE_KEY);
+        String stringSettings = node.getAttributes().get(REMOTE_STORE_SEGMENT_REPOSITORY_SETTINGS_ATTRIBUTE_KEY);
+
+        assert name != null : "Name cannot be null";
+        assert type != null : "Type cannot be null";
+        assert stringSettings != null : "Settings cannot be null";
+
+        return new RepositoryMetadata(name, type, buildSettings(stringSettings));
+
+    }
+
+    private static RepositoryMetadata buildTranslogRepositoryMetadata(DiscoveryNode node) {
+        String name = node.getAttributes().get(REMOTE_STORE_TRANSLOG_REPOSITORY_NAME_ATTRIBUTE_KEY);
+        String type = node.getAttributes().get(REMOTE_STORE_TRANSLOG_REPOSITORY_TYPE_ATTRIBUTE_KEY);
+        String stringSettings = node.getAttributes().get(REMOTE_STORE_TRANSLOG_REPOSITORY_SETTINGS_ATTRIBUTE_KEY);
+
+        assert name != null : "Name cannot be null";
+        assert type != null : "Type cannot be null";
+        assert stringSettings != null : "Settings cannot be null";
+
+        return new RepositoryMetadata(name, type, buildSettings(stringSettings));
+    }
+
+    /**
+     * Validated or adds the remote store repository to cluster state if it doesn't exist.
+     * @param joiningNode
+     * @param currentState
+     * @return updated cluster state
+     */
+    public static ClusterState validateOrAddRemoteStoreRepository(DiscoveryNode joiningNode, ClusterState currentState) {
+        List<DiscoveryNode> existingNodes = new ArrayList<>(currentState.getNodes().getNodes().values());
+
+        ClusterState newState = ClusterState.builder(currentState).build();
+
+        // TODO: Mutating cluster state like this can be dangerous, this will need refactoring.
+        if (existingNodes.size() == 0) {
+            // TODO: Will need to add the node to cluster state once the repository metadata is updated, such
```

Comment:
isn't new cluster state published with joining node added to cluster state in JoinTaskExecutor?

after new cluster state is published by active cluster manager, its not immediately persisted in local. The active cluster manager generates a new state, broadcasts a publish request to all nodes(including itself) and in response to that publish request active cluster manager persists the state. b/w publishing and persisting the new cluster state can a new nodes JoinTaskExecutor start?